core/src/banking_stage.rs

```diff
@@ -556,28 +562,33 @@ impl BankingStage {
         let mut consumed_buffered_packets_count = 0;
         let buffered_packet_batches_len = buffered_packet_batches.len();
         let mut proc_start = Measure::start("consume_buffered_process");
-        let mut reached_end_of_slot = None;
+        let mut reached_end_of_slot: Option<EndOfSlot> = None;
 
         RetainMut::retain_mut(
             buffered_packet_batches,
             |buffered_packet_batch_and_offsets| {
                 let (packet_batch, ref mut original_unprocessed_indexes, _forwarded) =
                     buffered_packet_batch_and_offsets;
-                if let Some((next_leader, bank)) = &reached_end_of_slot {
+                if let Some(end_of_slot) = &reached_end_of_slot {
                     // We've hit the end of this slot, no need to perform more processing,
                     // just filter the remaining packets for the invalid (e.g. too old) ones
-                    let new_unprocessed_indexes = Self::filter_unprocessed_packets(
-                        bank,
-                        packet_batch,
-                        original_unprocessed_indexes,
-                        my_pubkey,
-                        *next_leader,
-                        banking_stage_stats,
-                    );
-                    Self::update_buffered_packets_with_new_unprocessed(
-                        original_unprocessed_indexes,
-                        new_unprocessed_indexes,
-                    )
+                    // if the working_bank is available
+                    if let Some(bank) = &end_of_slot.working_bank {
+                        let new_unprocessed_indexes = Self::filter_unprocessed_packets(
+                            &bank,
+                            packet_batch,
+                            original_unprocessed_indexes,
+                            my_pubkey,
+                            end_of_slot.next_slot_leader,
+                            banking_stage_stats,
```

Comment:
So this filter will only run if we have a bank, i.e. we detected the end of the slot because we hit this case where transaction processing was interrupted due to the slot ending: https://github.com/solana-labs/solana/blob/26e160d4c43c58db167b8a753d2f002b0bc0f821/core/src/banking_stage.rs#L623-L625.

Is there anything essential here that needs to be done even if we were a bit late, the slot ended and the bank was cleared? Seems like this filter:

1. Filters out outdated transactions
2. Filters out stuff other than votes during vote only mode
3. Filters out duplicate transactions

If we think these are important, one thing we could do is pass the bank into `consume_buffered_packets`, since it's only called when we have a bank. We would just have to be careful to reset that bank to a new bank if in this loop we detected a new bank who's slot did not match the passed in bank though.

Reply:
> So this filter will only run if we have a bank, i.e. we detected the end of the slot because we hit this case where transaction processing was interrupted due to the slot ending:
> 
> https://github.com/solana-labs/solana/blob/26e160d4c43c58db167b8a753d2f002b0bc0f821/core/src/banking_stage.rs#L623-L625
> 
> .
> Is there anything essential here that needs to be done even if we were a bit late, the slot ended and the bank was cleared? Seems like this filter:
> 
> 1. Filters out outdated transactions
> 2. Filters out stuff other than votes during vote only mode
> 3. Filters out duplicate transactions
> 
> If we think these are important, one thing we could do is pass the bank into `consume_buffered_packets`, since it's only called when we have a bank. We would just have to be careful to reset that bank to a new bank if in this loop we detected a new bank who's slot did not match the passed in bank though.

This is basically current behavior - do nothing to packet_batch is no bank. But worth exam it; I think the first 2 tasks can be left till new bank becomes available - new bank has better idea about if TX is too old or if in vote-only mode; dedup perhaps can be done regardless bank status, but not sure how important it is to do here. So my instinct is to leave as is. 

If filtering when bank is gone is valid, could keep a copy of `working_bank` at Ln625 be a valid alternative of passing bank as param?

Reply:
> but not sure how important it is to do here. So my instinct is to leave as is.

Yup makes sense for now, thanks!